### ISSUE TITLE
Supporting PreTraining without streaming

### ISSUE THREAD

**Issue by TommasoBendinelli** [axolotl-ai-cloud/axolotl#2264]
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

Hi all, 

I was checking the code and I have seen the following:
```
def prepare_dataset(cfg, tokenizer, processor=None):
    prompters = []
    if not cfg.pretraining_dataset:
        with zero_first(is_local_main_process()):
            if cfg.test_datasets:
                train_dataset, _, prompters = load_prepare_datasets(
                    tokenizer,
                    cfg,
                    DEFAULT_DATASET_PREPARED_PATH,
                    split="train",
                    processor=processor,
                )
                _, eval_dataset, _ = load_prepare_datasets(
                    tokenizer,
                    cfg,
                    DEFAULT_DATASET_PREPARED_PATH,
                    split="test",
                    processor=processor,
                )
            else:
                train_dataset, eval_dataset, prompters = load_prepare_datasets(
                    tokenizer,
                    cfg,
                    DEFAULT_DATASET_PREPARED_PATH,
                    processor=processor,
                )
```
It looks like the dataset in  case of pre-training, the dataset is not loaded in advance; only streaming mode is supported. Is this expected? If so, how can I perform pre-training using a JSONL file?

### Current behaviour

If cfg.pretraining_dataset than a streaming dataset will be created

### Steps to reproduce

With a yml as follows:
pretraining_dataset:
  - path: /workspace/data/json_folder/
    name: default
    type: pretrain
    text_column: text

### Config yaml

```yaml

```

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.10

### axolotl branch-commit

0.6.0

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by NanoCode012** [comment]
Hello!

> It looks like the dataset in case of pre-training, the dataset is not loaded in advance; only streaming mode is supported. Is this expected?

Yes, `pretraining_dataset` was created to help facilitate streaming for large datasets.

> If so, how can I perform pre-training using a JSONL file?

If you want to do pre-training with JSONL, there are two ways.
1. Streaming: 
```yaml
pretraining_dataset:
  - path: json
    data_files: ["A.jsonl", "B.jsonl", "C.jsonl"]
```
2. Non-streaming:
```yaml
datasets:
  - path: A.jsonl
    type: completion
```

Non-streaming tokenizes in advance, whereas streaming tokenize on demand. Does this answer your question?

We're currently writing new docs on this to help clarify this better.

**Comment by TommasoBendinelli** [comment]
Thanks a lot! so if I understand correctly, type "completion" is equivalent to pretraining_dataset without streaming

**Comment by NanoCode012** [comment]
Yes, that's right

**Comment by TommasoBendinelli** [comment]
Alright, thanks! Then we can close this. I recommend adding this to the documentation.
